File: src/domain/models/email.rs
const FORBIDDEN_DOMAINS: [&str; 13] = [
    "gmail",
    "yahoo",
    "hotmail",
    "aol",
    "wanadoo",
    "msn",
    "live",
    "free",
    "outlook",
    "laposte",
    "protonmail",
    "yopmail",
    "minutemail",
];

#[derive(Clone, Debug, PartialEq)]
pub struct Email {
    raw: String,
}

impl Email {
    #[cfg(test)]
    pub fn test() -> Self {
        Self {
            raw: String::from("bonjour@salaires.dev"),
        }
    }
}

#[derive(Clone, Debug)]
pub enum Error {
    NotContainingAnAt,
    NotPro,
}

impl TryFrom<String> for Email {
    type Error = Error;

    fn try_from(raw: String) -> Result<Self, Self::Error> {
        if raw.contains("@") {
            if FORBIDDEN_DOMAINS
                .iter()
                .any(|domain| raw.as_str().contains(domain))
            {
                Err(Error::NotPro)
            } else {
                Ok(Self { raw })
            }
        } else {
            Err(Error::NotContainingAnAt)
        }
    }
}

impl From<Email> for String {
    fn from(email: Email) -> Self {
        email.raw
    }
}

File: src/domain/models/email.ts
import { Result } from "@utils";

export type Email = {
    _type: "email";
    raw: string;
};

export type EmailError =
    | "not_an_email"
    | "not_a_pro_email"
    ;

const FORBIDDEN_DOMAINS = [
    "gmail",
    "yahoo",
    "hotmail",
    "aol",
    "wanadoo",
    "msn",
    "live",
    "free",
    "outlook",
    "laposte",
    "protonmail",
    "yopmail",
    "minutemail",
];

export const Email = {
    generate(): Email {
        return {
            _type: "email",
            raw: "bonjour@salaires.dev"
        };
    },
    toString(email: Email): string {
        return email.raw;
    },
    tryFromString(email: string, { admin = false } = {}): Result<Email, EmailError> {
        const raw = email.trim();
        if (! raw.includes("@")) { return Result.err("not_an_email"); }
        if (admin) { return Result.ok({ _type: "email", raw }); }
        if (FORBIDDEN_DOMAINS.some(domain => raw.includes(domain))) { return Result.err("not_a_pro_email"); }
        return Result.ok({ _type: "email", raw });
    }
};
